Service key support

gberche-orange opened the pull request:
This PR add ability for ServiceInstanceAppBindingWorkflow impls to use service keys to expose some backing services credentials (see #285 for more background about related use-case). It includes CT and AT.

Other improvements were added to the AT:
- documentation enhancements
- logging deployed broker recents logs

gberche-orange commented:
Started rebasing and fixing new PMD and checkstyle rules from #291 Still some checkstyle fixes to be brought

Original PR before is available at https://github.com/orange-cloudfoundry/osb-cmdb-spike/tree/service-key-support-before-spring-style-rebase

Albertoimpl commented:
Thanks for the contribution, would you mind rebasing this PR @gberche-orange? Thanks!

gberche-orange commented:
@Albertoimpl the PR was rebased against master and adapted to new checkstyles and pmd rules. I'm looking forward to your review.

Albertoimpl commented on spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/deployer/BackingServicesProvisionService.java:
```diff
@@ -28,4 +29,15 @@
 
+	/**
+	 * Create service keys
+	 * @return A flux of credentials associated with each created service key
```
Please, delete comment to be consistent with the other methods.

gberche-orange replied:
As the content of the returned value is assigned deep in CloudFoundryAppDeployer, and felt not that intuitive obvious to newcomers, I rather added javadoc on all methods of the class as to still preserve class consistency. Please let me know if you have a better suggestion on how to remove ambiguity to code readers on the intent and content of the returned type.

Albertoimpl commented on spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/deployer/BackingServicesProvisionService.java:
```diff
@@ -28,4 +29,15 @@
+
+	/**
+	 * Delete service keys
+	 * @return a Flux of service key names that were deleted
```
Please, delete comment to be consistent with the other methods.

gberche-orange replied:
Same as https://github.com/spring-cloud/spring-cloud-app-broker/pull/292/files#r392504267

Albertoimpl commented on spring-cloud-app-broker-deployer/src/main/java/org/springframework/cloud/appbroker/deployer/CreateServiceKeyRequest.java:
```diff
@@ -0,0 +1,113 @@
+
+	private final Map<String, Object> parameters;
+
+	private final Map<String, String> properties;  // to access space target property
```
Please, delete comment to be consistent with the other methods.

gberche-orange replied:
fixed in 10fd24ea9ebc1ce01db90c1eee8a3e2a089fb46d

Albertoimpl commented on spring-cloud-app-broker-deployer/src/main/java/org/springframework/cloud/appbroker/deployer/CreateServiceKeyResponse.java:
```diff
@@ -0,0 +1,68 @@
+public class CreateServiceKeyResponse {
+
+	/**
+	 * Service key name
```
Please, delete comment to be consistent with the other fields.

gberche-orange replied:
fixed in 10fd24ea9ebc1ce01db90c1eee8a3e2a089fb46d

Albertoimpl commented on spring-cloud-app-broker-deployer/src/main/java/org/springframework/cloud/appbroker/deployer/DeleteServiceKeyRequest.java:
```diff
@@ -0,0 +1,79 @@
+
+	private final String serviceInstanceName;
+	private final String serviceKeyName;
+	private final Map<String, String> properties; // to access space target property
```
Please, delete comment to be consistent with the other fields.

gberche-orange replied:
fixed in 10fd24ea9ebc1ce01db90c1eee8a3e2a089fb46d

Albertoimpl commented:
Thanks a lot for the contribution! I don't see any problem with the implementation nor with the way it is implemented, apart from the lack of ATs and CTs that should be included and the picky notes about the comments I left. 

There are a couple of things that do not feel right but I am happy to re-evaluate them in a future release when we start implementing the k8s support.
- You can achieve this behaviour by just implementing a `CreateServiceInstanceWorkflow`.
- In a k8s world, there is not a direct mapping to a Service Key, so leaking that name, that is only CF-related, as part of the `AppDeployer` interface, which is a generic one, doesn't feel right.

As said, if you add the ATs and CTs, I am happy with including it. Thanks a lot for the great work!

gberche-orange commented:
Thanks @Albertoimpl for your review. I'll try working on providing the CT and AT now that I see there is interest.

> You can achieve this behaviour by just implementing a CreateServiceInstanceWorkflow.

Actually, I'm currently using the service key support in a custom `CreateServiceInstanceAppBindingWorkflow`, see [osbcmdb/AppDeploymentCreateServiceBindingWorkflow.java](https://github.com/orange-cloudfoundry/osb-cmdb-spike/blob/08390cef44622db08e11adc95ae8d4b2110bc6a2/osb-cmdb/src/main/java/com/orange/oss/osbcmdb/AppDeploymentCreateServiceBindingWorkflow.java#L74-L100). I first considered implementing this without SCAB support, but this required to duplicate some of the existing beans such as `DefaultBackingServicesProvisionService` and `CloudFoundryAppDeployer`, so I'm happy if this can make it into SCAB and be useful to others.

> In a k8s world, there is not a direct mapping to a Service Key, so leaking that name, that is only CF-related, as part of the AppDeployer interface, which is a generic one, doesn't feel right.

Yes. I understand that the service key could map to a K8S service cat `ServiceBinding` and fetching credentials from the associated `Secret`, see https://kubernetes.io/docs/concepts/extend-kubernetes/service-catalog/#mapping-the-connection-credentials `These pieces of information are stored in secrets that the application in the cluster can access and use to connect directly with the managed service.`

So potentially, a K8S-deployer implementation of the service-key methods could be brought in the future. I do agree that then the name "ServiceKey" could be more generic such as "ServiceCredentials".

Albertoimpl commented:
Great to hear!
> Thanks @Albertoimpl for your review. I'll try working on providing the CT and AT now that I see there is interest.

Looking forward to reviewing them. I typically start working always with an AT, then a CT and then I start coding the rest.

> DefaultBackingServicesProvisionService and CloudFoundryAppDeployer, so I'm happy if this can make it into SCAB and be useful to others.

Perfect! Thanks for contributing. I am sure this will benefit others.

> So potentially, a K8S-deployer implementation of the service-key methods could be brought in the future. I do agree that then the name "ServiceKey" could be more generic such as "ServiceCredentials".

It would be ideal if you made this rename.

Again, thanks a lot for your contributions!

gberche-orange commented:
@Albertoimpl Sorry for long silence on this PR. I've rebased the PR and added component test covering service key usage, and addressed suggested changes in the review. Can you please have a new look ?

Note that I'm struggling with checkstyle import rule on https://github.com/orange-cloudfoundry/osb-cmdb-spike/blob/5016f4f5b1bd28d7ba9066bdfe3bc766b8ce397f/spring-cloud-app-broker-integration-tests/src/test/java/org.springframework.cloud.appbroker/integration/CreateBindingWithServiceKeyComponentTest.java#L54
* Besides, gradle checkstyle fails on my development box  with symptom
`Unable to create Root Module: config {.../src/checkstyle/checkstyle.xml}`, which looks similar to https://stackoverflow.com/questions/49645147/checkstyle-unable-to-create-root-module which I need to pursue. Any hint on fixing this would help, as waiting for circleci builds feedback implies a 15 min feedback time.
* I tried to follow the [CONTRIBUTING.adoc](https://github.com/spring-cloud/spring-cloud-app-broker/blob/master/CONTRIBUTING.adoc) which recommends usage of the eclipse code formatter, and applied the intellij "optimize import" action. Is there other automated import formatting that the app broker team uses within intellij ?

I'll now try to provide an acceptance test. Is there a way I would be able see test results associated with the PR for acceptance tests (such as making the associated concourse pipeline public or adding 3rd party contributors as viewers) ?

gberche-orange commented:
@Albertoimpl Re on acceptance test covering service key support. Such test would probably ask for a service key on the brokered service instance (i.e. an OSB service binding call with an empty appId in the bindResource), hence requiring a service binding workflow.

Would that make sense to make this workflow available in core (for integration-tests and acceptance-tests) to use, and leave it up to app broker users to enable it, for instance using a property ?

gberche-orange commented:
Identified steps for contributing the AT
* [ ] set up the local environment to run the test
   * The [documentation](https://github.com/spring-cloud/spring-cloud-app-broker/blob/283819aeecca8c4e7f48ba4b661116b8c31aa2ce/spring-cloud-app-broker-acceptance-tests/README.adoc) is missing the required authorities for the UAA client
* [ ] execute existing test cases
* [ ] develop new test
   * [ ] handle clean up of provisioned element on test failures
      * [ ] reverse engineer and document element provisionned
      * [ ] document clean up script manually invoked
      * [ ] improve the `CloudFoundryAcceptanceTest` to be resilient to failed/interrupted tests that did not execute cleanup
* [ ] execute the new test to get test status
   * [ ] import concourse ci pipeline

Albertoimpl commented:
Thanks a lot for the contributions, it is looking great!
We added a different style guide than the documented in `CONTRIBUTING`, sorry about that, we are going to fix it: https://github.com/spring-cloud/spring-cloud-app-broker/issues/345 and you can find all the check styles over here: https://github.com/spring-cloud/spring-cloud-app-broker/tree/master/src

Regarding the ATs, we have them ready to be public, we need to move the pipeline to a public concourse instance: https://github.com/spring-cloud/spring-cloud-app-broker/issues/346 
However, you should be able to run them in your own environments. Let me know if you find any issues with that.

Thanks!

gberche-orange commented:
@Albertoimpl I've updated the PR with 
* service instance deprovision revert (fixes #344)
* acceptance test for service key support which passes on my environment.

The circle ci build is still failing on checkstyle/pmd violations that I'm trying to fix. 

Can you please resume the review of this PR when possible ?

gberche-orange commented:
@Albertoimpl @royclarkson @LittleBaiBai As a less intrusive fix for #344 than 
b2324e4 and 068db25, I'm now considering instead to add distinct workflows only triggering for brokered services with only backing services, e.g. `ServicesOnlyDeploymentDeleteServiceInstanceWorkflow` and `ServicesOnlyDeploymentDeleteServiceInstanceWorkflow` with the `accept()` method selecting the right workflow depending on the absence of backing apps.

This should enable both behaviors to co-exist without negatively impacting each other, while sharing with the app broker authors the service key support (which @Albertoimpl found valuable in https://github.com/spring-cloud/spring-cloud-app-broker/pull/292#issuecomment-567013071 )

Please let me know if you have other ideas.

I hope this will help review/merge of this PR, and also convince my team that I'm heading in the right direction when I invest time and effort to contribute to back app broker while developing features for [orange-cloudfoundry/osb-cmdb-spike](https://github.com/orange-cloudfoundry/osb-cmdb-spike)

gberche-orange commented:
This [osb-cmdb](https://github.com/orange-cloudfoundry/osb-cmdb) related feature seems to not make much sense into app-broker use-cases.  Closing it.

Thanks however @Albertoimpl for your past review and support on this PR !
